clab/general.go

```diff
@@ -0,0 +1,126 @@
+package clab
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/cloudflare/cfssl/log"
+	"github.com/srl-labs/containerlab/types"
+)
+
+var (
+	CLAB_HOSTENTRY_PREFIX  = "###### CLAB-%s-START ######"
+	CLAB_HOSTENTRY_POSTFIX = "###### CLAB-%s-END ######"
+)
+
+func AppendHostsFileEntries(containers []types.GenericContainer, labname string) error {
+	if labname == "" {
+		return fmt.Errorf("missing lab name")
+	}
+	// lets make sure we do not have remaining of a non destroyed run in the hosts file
+	err := DeleteEntriesFromHostsFile(labname)
+	if err != nil {
+		return err
+	}
+	data := GenerateHostsEntries(containers, labname)
+	if len(data) == 0 {
+		return nil
+	}
+	f, err := os.OpenFile("/etc/hosts", os.O_APPEND|os.O_WRONLY, os.ModeAppend)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	_, err = f.WriteString("\n")
+	if err != nil {
+		return err
+	}
+	_, err = f.Write(data)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// hostEntries builds an /etc/hosts compliant text blob (as []byte]) for containers ipv4/6 address<->name pairs
+func GenerateHostsEntries(containers []types.GenericContainer, labname string) []byte {
+	entries := strings.Builder{}
+	v6entries := strings.Builder{}
+
+	entries.WriteString(fmt.Sprintf(CLAB_HOSTENTRY_PREFIX+"\n", labname))
+
+	for _, cont := range containers {
+		if len(cont.Names) == 0 {
+			continue
+		}
+		if cont.NetworkSettings.Set {
+			if cont.NetworkSettings.IPv4addr != "" {
+				fmt.Fprintf(&entries, "%s\t%s\n", cont.NetworkSettings.IPv4addr, strings.TrimLeft(cont.Names[0], "/"))
+			}
+			if cont.NetworkSettings.IPv6addr != "" {
+				fmt.Fprintf(&v6entries, "%s\t%s\n", cont.NetworkSettings.IPv6addr, strings.TrimLeft(cont.Names[0], "/"))
+			}
+		}
+	}
+
+	entries.WriteString(v6entries.String())
+	entries.WriteString(fmt.Sprintf(CLAB_HOSTENTRY_POSTFIX+"\n", labname))
+	return []byte(entries.String())
+}
```

Comment:
Overall I'd lean toward `*printf`, but might be a personal preference.

See this can likely also be done with `bytes.Buffer` 

The grouping is good, especially since the target here is human readability with the comment block

Reply:
- Iterating twice vs using 2 io.Writer is arguable, but fair enough let's use 2 writers.
- The reason the method returns a []byte is because the output will be written to a file as []byte. 
`strings.Builder` uses a []byte under the hood, it was added to Go as a better alternative to `fmt.Sprintf` if the user knows the items to be written are strings and has the intention of using the result as a `string` (that's why strings.Builder doesn't have/need a `.Bytes()` method but `bytes.Buffer` does).
The use of `strings.Builder` in this case, writes strings to bytes (in the builder) then converts them to a string to then convert the whole thing to []byte to be returned.
If we return a string, the function writing to the file will have to covert to []byte anyways.
- Using string formatters `%s` implies that `fmt.Sprintf` has to do a type reflection on the received `interface{}` to determine its type and find its string representation. which defeats the purpose of buffered writers (`strings.Builder` and/or `bytes.Buffer`)

Reply:
OK, I think that is all now as discussed separately with you @karimra
I'll set this to resolved